File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofServices/PlatformStatsService.swift
import Foundation
import Supabase
import WagerproofModels

/// Reads for the Agents "Platform Statistics" screen. Population-level analytics
/// (every agent's win-rate / net-units distribution) — kept separate from the
/// per-agent / leaderboard reads in `AgentPerformanceService`.
///
/// Both calls hit whole-population / public RPCs added in
/// `20260701120000_agent_performance_distribution_rpc.sql`. The distribution
/// fetch returns raw per-agent rows once; all the screen's controls (threshold
/// slider, bin width, metric, sport) re-bucket in memory with no refetch.
public enum PlatformStatsService {
    /// Every agent with at least `minDecided` settled picks (wins+losses).
    /// `minDecided: 1` (the default) drops pending-only agents server-side; the
    /// interactive ≥N threshold is applied client-side on top of these rows.
    public static func fetchAgentDistribution(minDecided: Int = 1) async throws -> [AgentStatDatum] {
        let main = await MainSupabase.shared.client
        struct Params: Encodable { let p_min_decided: Int }
        let rows: [AgentStatDatum] = try await main
            .rpc("get_agent_performance_distribution", params: Params(p_min_decided: minDecided))
            .execute()
            .value
        return rows
    }

    /// Top public agents whose `metric` falls in `[lower, upper]`, ranked by net
    /// units, each with their currently-open (pending) picks — the tap-a-bar
    /// drill-down. `sport` scopes the returned pending picks (nil = all sports).
    public static func fetchBinAgents(
        metric: StatMetric,
        sport: AgentSport? = nil,
        lower: Double,
        upper: Double,
        minDecided: Int,
        limit: Int = 20
    ) async throws -> [BinAgent] {
        let main = await MainSupabase.shared.client
        struct Params: Encodable {
            let p_metric: String
            let p_sport: String?
            let p_lower: Double
            let p_upper: Double
            let p_min_decided: Int
            let p_limit: Int
        }
        let params = Params(
            p_metric: metric.sqlValue,
            p_sport: sport?.rawValue,
            p_lower: lower,
            p_upper: upper,
            p_min_decided: minDecided,
            p_limit: limit
        )
        let rows: [BinAgent] = try await main
            .rpc("get_distribution_bin_agents", params: params)
            .execute()
            .value
        return rows
    }
}

private extension StatMetric {
    /// The `p_metric` value the RPC expects (`avatar_performance_cache` column
    /// naming) — distinct from the Swift enum's rawValue.
    var sqlValue: String {
        switch self {
        case .winRate: return "win_rate"
        case .netUnits: return "net_units"
        }
    }
}

File: wagerproof-android-native/core/services/src/main/kotlin/com/wagerproof/core/services/PlatformStatsService.kt
package com.wagerproof.core.services

import com.wagerproof.core.models.AgentSport
import com.wagerproof.core.models.AgentStatDatum
import com.wagerproof.core.models.BinAgent
import com.wagerproof.core.models.StatMetric
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.rpc
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Population-level analytics for the Agents "Platform Statistics" screen (Main
 * project) — kept separate from the per-agent/leaderboard reads in
 * [AgentPerformanceService]. Port of iOS `PlatformStatsService.swift`.
 */
object PlatformStatsService {

    /**
     * Every agent with at least [minDecided] settled picks (wins+losses).
     * Raw per-agent rows fetched once; all re-bucketing (threshold slider,
     * bin width, metric, sport) happens client-side with no refetch.
     */
    suspend fun fetchAgentDistribution(minDecided: Int = 1): List<AgentStatDatum> =
        SupabaseClients.main.postgrest.rpc(
            "get_agent_performance_distribution",
            buildJsonObject { put("p_min_decided", minDecided) },
        ).decodeList()

    /**
     * Top public agents whose [metric] falls in `[lower, upper]`, ranked by net
     * units, each with their currently-open picks — the tap-a-bar drill-down.
     * [sport] scopes the returned pending picks (null = all sports).
     */
    suspend fun fetchBinAgents(
        metric: StatMetric,
        sport: AgentSport? = null,
        lower: Double,
        upper: Double,
        minDecided: Int,
        limit: Int = 20,
    ): List<BinAgent> {
        val params = buildJsonObject {
            put("p_metric", metric.sqlValue)
            sport?.let { put("p_sport", it.raw) }
            put("p_lower", lower)
            put("p_upper", upper)
            put("p_min_decided", minDecided)
            put("p_limit", limit)
        }
        return SupabaseClients.main.postgrest
            .rpc("get_distribution_bin_agents", params)
            .decodeList()
    }
}

// The p_metric value the RPC expects (`avatar_performance_cache` column
// naming) — distinct from StatMetric.raw, which is the client-side key.
private val StatMetric.sqlValue: String
    get() = when (this) {
        StatMetric.WIN_RATE -> "win_rate"
        StatMetric.NET_UNITS -> "net_units"
    }
